packages/kbn-expandable-flyout/src/hooks/use_url_state.ts

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { useCallback, useEffect, useState } from 'react';
+import { encode, decode, RisonValue } from '@kbn/rison';
+
+interface StateCache {
+  namespaces: Record<string, Record<string, RisonValue | undefined>>;
+  timeout: number;
+}
+
+const cache: StateCache = {
+  namespaces: {},
+  timeout: 0,
+};
+
+export const useUrlState = (rootKey: string, key: string) => {
```

Comment:
let's add some documentation here as well as in the code. It's not super straightforward to see at first glance what the listener exactly does for example